interpreter/mesh/src/main/scala/io/buoyant/interpreter/mesh/Client.scala

```diff
@@ -0,0 +1,305 @@
+    resolver: mesh.Resolver,
+    delegator: mesh.Delegator
+  ) extends NameInterpreter with Delegator {
+
+    /**
+     * When observed, streams bound trees from the mesh Interpreter.
+     *
+     * Leaf nodes include a Var[Addr] that, when observed, streams
+     * replica resolutions from the mesh Resolver.
+     */
+    override def bind(dtab: Dtab, path: Path): Activity[NameTree[Name.Bound]] = {
+      val open = () => interpreter.streamBoundTree(mkBindReq(namespace, path, dtab))
+      streamActivity(open, decodeBoundTree)
+    }
+
+    /**
+     * When observed, streams dtabs from the mesh server and
+     */
+    override lazy val dtab: Activity[Dtab] = {
+      val open = () => delegator.streamDtab(mkDtabReq(namespace))
+      streamActivity(open, decodeDtab)
+    }
+
+    override def delegate(
+      dtab: Dtab,
+      tree: NameTree[Name.Path]
+    ): Activity[DelegateTree[Name.Bound]] = {
+      val open = () => delegator.streamDelegateTree(mkDelegateTreeReq(namespace, dtab, tree))
+      streamActivity(open, decodeDelegateTree)
+    }
+
+    private[this] val resolve: Path => Var[Addr] = {
+      case Path.empty => Var.value(Addr.Failed("empty"))
+      case id =>
+        val open = () => resolver.streamReplicas(mkReplicasReq(id))
+        streamVar(Addr.Pending, open, replicasToAddr)
+    }
+
+    private[this] val fromBoundNameTree: mesh.BoundNameTree => NameTree[Name.Bound] =
+      mkFromBoundNameTree(resolve)
+
+    private[this] val decodeBoundTree: mesh.BoundTreeRsp => Option[NameTree[Name.Bound]] = {
+      case mesh.BoundTreeRsp(Some(ptree)) => Some(fromBoundNameTree(ptree))
+      case _ => None
+    }
+
+    private[this] val fromBoundDelegateTree: mesh.BoundDelegateTree => DelegateTree[Name.Bound] =
+      mkFromBoundDelegateTree(resolve)
+
+    private[this] val decodeDelegateTree: mesh.DelegateTreeRsp => Option[DelegateTree[Name.Bound]] = {
+      case mesh.DelegateTreeRsp(Some(ptree)) => Some(fromBoundDelegateTree(ptree))
+      case _ => None
+    }
+  }
+
+  private[this] val releaseNop = () => Future.Unit
+
+  /**
+   * Constructs a Var that, when observed, calls `open()` to obtain a
+   * gRPC stream, publishing updates into the Var's state.
+   *
+   * When the stream returns an error, the Var may be updated (as
+   * according to `toT`) and no further updates will will be
+   * published.
+   */
+  private[this] def streamVar[S, T](
+    init: T,
+    open: () => Stream[S],
+    toT: Try[S] => Option[T]
+  ): Var[T] = Var.async[T](init) { state =>
+    val rsps = open()
+
+    // As we receive streamed messages, we are careful not to release
+    // them until the state has been updated to a new value. This is
+    // intended to (1) integrate tightly with flow control and more
+    // importantly (2) later integrate with netty's reference counted
+    // bueffers.
+    @volatile var closed = false
+    def loop(releasePrior: () => Future[Unit]): Future[Unit] =
+      if (closed) Future.Unit
```

Comment:
If the Var becomes unobserved, the prior message will never be released, even if the Var becomes observed again.  I think this is fine because when the Var is observed again a new stream will be created.  Will the old stream be torn down and collected if it has unreleased frames on it?